2015/presentations/sztuczki_z_metoda_super-Dominik-Socek-Dlugajczyk.md

```diff
@@ -0,0 +1,64 @@
+
+Niestety taka sztuczka ma sens tylko w testach, gdyż wtedy wiemy dokładnie jaką klasę blokujemy i jak MRO się zachowa (oraz jak struktura się zmienii). Jeśli byśmy chcieli jeszcze raz podziedziczyć po wielu klasach, to hierarchia MRO może ulec zmianie i nasz kod przestanie działać. Tak samo nie możemy zablokować kodu konkretnej klasy w hierarchii, już nie mówiąc o tym, że trzeba by przepisać wszystkie te klasy, po których nasza klasa będzie dziedziczona, aby zmienić strukturę dziedziczenia (co jest niewykonalne, gdyż nigdy nie wiemy co będzie dziedziczyło po naszej klasie).
+
+Wiedząc to wszystko bardzo łatwo będzie nam teraz zrozumieć po co w argumentach metody super trzeba podać aktualną klasę oraz aktualny obiekt. Czyli:
+
+```super(OrdersListController, self)```
+
+Pierwszym argument jest miejsce w liście MRO, drugie jest to obiekt od której lista się zaczyna. Na przykład: "self" może być typu OrdersListController, jeśli stworzyliśmy obiekt tej klasy, natomiast jeśli zrobiliśmy klasę która tylko dziedziczy po OrdersListController, to self będzie typu tej nowej klasy. Dlatego użycie super w tej formie nie ma sensu:
+```super(self.__class__, OrderListController)```
+
+Zauważyłem jakiś czas temu, że bardzo często używam super() na początku każdej metody. Chciałem zatem zrobić dekorator @superme, który by wykonywał super za mnie. Nie udało mi się to, gdyż dekoratory działają w dość prosty sposób. Dekoratory jest to funkcja, która bierze funkcję i zwraca inną funkcję. Jeśli zrobimy dekorator na metodzie, to w klasie wyciągniemy tę metodę i wystawimy nową. Jeśli spróbujemy w tej nowej funkcji użyć super, to wszystko powinno grać. Jeśli natomiast spróbujemy użyć super na metodzie, której rodzic ma metodę pod dekoratorem, to używając super nie dostaniemy naszej upragnionej metody, lecz funkcję zwróconą przez dekorator. Czyli mając coś takiego:
+
+```
+def superme(fun):
+    def wrapper(*args, **kwargs):
+        return fun(*args, **kwargs)
+    return wrapper
+
+
+class One(object):
+
+    @superme
+    def method(self):
+        pass
+
+
+class Two(One):
+
+    def method(self):
+        print(super().method)
+
+Two().method()
+```
```

Comment:
To było testowane na python3 gdzie ten kod jest poprawny.
Poprawiłem tak, aby działało i na python2 i na python3.